lib/logstash/filters/elasticsearch.rb

```diff
@@ -191,4 +192,55 @@ def extract_total_from_hits(hits)
   def test_connection!
     get_client.client.ping
   end
+
+  private
+
+  ##
+  # Map the provided array-of-strings to an array of `URI::Generic`
+  # instances, which the Elasticsearch client can use to establish
+  # connections.
+  #
+  # @param hosts [Array<String>]: one or more hosts, each is one of:
+  #                               - a bare hostname or ip, optionally
+  #                                 followed by a colon and port number
+  #                               - a square-bracketed ipv6 literal,
+  #                                 optionally followed by a colon and port
+  #                                 number
+  #                               - a qualified URL with http/https schema
+  # @param force_ssl [Boolean]: true to force SSL; will cause failure if one
+  #                             or more hosts explicitly supplies non-SSL
+  #                             scheme (e.g., `http`).
+  #
+  # @return [Array<URI::Generic>]
+  def normalise_hosts(hosts, force_ssl)
+    hosts.map do |input|
+      if force_ssl && input.start_with?('http://')
+        logger.error("Plugin configured to force SSL with `ssl => true`, " +
+                     "but a host explicitly declared non-https URL `#{input}`")
+
+        raise LogStash::ConfigurationError, "Aborting due to conflicting configuration"
+      end
+
+      if input.start_with?('http://','https://')
+        URI::Generic.new(*URI.split(input))
```

Comment:
I suggest we document why this remarshalling is necessary here, it will seem odd for fresh eyes.

Reply:
Added a comment to add clarity. We're not actually remarshalling, this is the only way using public APIs to take a string input and create an `URI::Generic`.